### PR TITLE
viewpager2

### PR DESCRIPTION
**Pull request by troZee**
# Summary

I migrated whole ViewPager  to version 2. All provided api should work after migration. 
This is alpha version. I needed to copy and paste whole VP2 from the google source, because there was a problem with RecyclerView and I needed to fix it.  It means, that we need to maintenance it on our own once we want to bump the version of VP2.  

Close https://github.com/react-native-community/react-native-viewpager/issues/118

## Test Plan
* Check example
* Check this migration in your existing app

I did some tests on: 
* Samsung S8 
* Moto X4 
* Xiaomi RedMi 5A

And all API works fine without any breaks

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ❌     |
| Android |    ✅     |

## Checklist

- [x] I have tested this on a device and a simulator
- [ ] I added the documentation in `README.md`
- [ ] I mentioned this change in `CHANGELOG.md`

### CODE REVIEW

**Comment by Krizzu** [comment]
What kind of an issue you had with RV that you had to copy over the code?

**Comment by troZee** [comment]
> What kind of an issue you had with RV that you had to copy over the code?

Set page without animation methods does not work properly due : 

https://stackoverflow.com/questions/49371866/recyclerview-wont-update-child-until-i-scroll/49372697#49372697

The problem is, that there is no patch-package for gradle, so I needed to copy and paste whole code from google's repo.

**Comment by osdnk** [comment]
Probably I'm not getting what's the case. Could you explain it more deeply?

**Comment by troZee** [comment]
I was struggling with `setCurrentItem` method, because it did not go to specific page:



As you can see once I am clicking on next button, page does not change. I was digging to find solution of this problem and tried [this](https://stackoverflow.com/questions/49371866/recyclerview-wont-update-child-until-i-scroll/49372697#49372697). 
 
Above [patch](https://stackoverflow.com/questions/49371866/recyclerview-wont-update-child-until-i-scroll/49372697#49372697) resolved above problem as follow:

https://cdn.discordapp.com/attachments/575311183897296906/672797486338932796/untitled2222.gif

TBH, I don't know, what Recycler View does under the hood and why this change works. I assume , that RV performs `scrollToPosition` method, which uses `requestLayout`, which somehow does not refresh the view, so we need to force this measuring again.

**Comment by troZee** [comment]
I know that there is lots of files, which has been added to PR, but please check only those: 
- `ViewPagerFragment.java`
- `ReactViewPagerManager.java`
- `FragmentAdapter.java`
- `*.js`

**Comment by ferrannp** [comment]
I agree that if we need to get all the source code, better to have our own copy in maven.

**Comment by Kasemyassine2** [comment]
I've tried your repo @troZee there was an issue with the padding around the component as well it was crashing while scrolling and pressing buttons while scrolling. I decided to stay on the official version.

What I have is a reading app with horizontal viewpager in which contains pages of articles which can be liked followed opened ... so it didnt work in my case as expected I just wanted to report those since they were fixed in previous releases .

**Comment by troZee** [comment]
@Kasemyassine2 thank you for the feedback. Once I publish VP2 patch package, we will release VP2 as RC on separate branch. Then we will create a special issue for this RC, because it is hard to developing and fixing at the one time. I would like to focus on deliver VP2. It would be simpler and cleaner, if everything will be as separate issue (Because I need repo to reproduce etc.).

File: android/src/main/java/com/reactnativecommunity/viewpager/ReactViewPagerManager.java [hunk 1]
```diff
@@ -1,132 +1,197 @@
 /**
  * Copyright (c) Facebook, Inc. and its affiliates.
- *
+ * <p>
```
Comment (Krizzu) [inline android/src/main/java/com/reactnativecommunity/viewpager/ReactViewPagerManager.java]: you don't need this

File: android/.classpath [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Krizzu) [inline android/.classpath]: you probably don't need this?

File: android/.settings/org.eclipse.buildship.core.prefs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+arguments=
```
Comment (Krizzu) [inline android/.settings/org.eclipse.buildship.core.prefs]: you probably don't need this, looks like auto generated setting file

File: android/src/main/java/com/reactnativecommunity/viewpager/ViewPagerFragment.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        ViewPagerFragment fragment = new ViewPagerFragment();
+        fragment.setArguments(args);
+        return fragment;
+    }
+
+
+    @Nullable
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
+        super.onCreateView(inflater, container, savedInstanceState);
```
Comment (Krizzu) [inline android/src/main/java/com/reactnativecommunity/viewpager/ViewPagerFragment.java]: you don't need to call super here

File: example/android/.settings/org.eclipse.buildship.core.prefs [hunk 5]
```diff
@@ -0,0 +1,2 @@
+connection.project.dir=
```
Comment (Krizzu) [inline example/android/.settings/org.eclipse.buildship.core.prefs]: can be deleted

File: example/android/app/.settings/org.eclipse.buildship.core.prefs [hunk 6]
```diff
@@ -0,0 +1,2 @@
+connection.project.dir=..
```
Comment (Krizzu) [inline example/android/app/.settings/org.eclipse.buildship.core.prefs]: sam here - can be deleted

**Comment by troZee** [comment]
`COMMAND_SET_SCROLL_ENABLED` has been removed. Please add support for this method @troZee

**Comment by malacca** [comment]
@troZee  debug, hot refresh

> java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.View.unFocus(android.view.View)' on a null object reference

**Comment by miguelocarvajal** [comment]
I am experiencing the issue where an embedded ScrollView will not work in a pager on Android.

So I wanted to test this ViewPager2 version to see if it solves it and provide feedback.

However I get a "Could not find com.github.troZee:ViewPager2:v1.0.6" error.

I added
``` "@react-native-community/react-native-viewpager": "https://github.com/react-native-community/react-native-viewpager.git#@troZee/viewpager2",```
to package.json and then ran "yarn run android".

What am I doing wrong?

**Comment by troZee** [comment]
hey @miguelocarvajal , thank you for reporting an issue. I can see, this repo still exist in jitpack. https://jitpack.io/#troZee/ViewPager2

Have you run `yarn install` once you added above line ?

**Comment by miguelocarvajal** [comment]
Definitely did run ```yarn install```, also tried manually removing the line from package.json and then re-adding it via ```yarn add```.

Both ways got the same error message described.

**Comment by miguelocarvajal** [comment]
Hey @troZee 

Got this version working.. it was a typo that was preventing it from working.

I can verify that a ScrollView in a vertical pager does not work correctly.

Here is a little component to reproduce the problem:

```
import ViewPager from "@react-native-community/viewpager";
import React from "react";
import { ScrollView, StyleSheet, Text, View } from "react-native";

const styles = StyleSheet.create({
  viewPager: {
    flex: 1,
  },
});

const App = () => (
  <ViewPager orientation="vertical" style={styles.viewPager} initialPage={0}>
    <View style={{ padding: 50, backgroundColor: "blue" }}>
      <Text>Vertical pager, go down to see problem</Text>
    </View>

    <View style={{ padding: 50, backgroundColor: "red" }}>
      <Text>ScrollView in vertical pager does not work</Text>

      <ScrollView>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
        <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
          Line
        </Text>
      </ScrollView>
    </View>
  </ViewPager>
);

export default App;
```

File: android/src/main/java/com/reactnativecommunity/viewpager/ReactViewPagerManager.java [hunk 7]
```diff
@@ -1,132 +1,204 @@
...
+
+
+    @ReactProp(name = "scrollEnabled", defaultBoolean = true)
+    public void setScrollEnabled(ViewPager2 viewPager, boolean value) {
+        viewPager.setUserInputEnabled(value);
+    }
+
+    @ReactProp(name = "orientation")
+    public void setOrientation(ViewPager2 viewPager, String value) {
+        viewPager.setOrientation(value.equals("horizontal") ? ORIENTATION_HORIZONTAL : ViewPager2.ORIENTATION_VERTICAL);
```
Comment (ferrannp) [inline android/src/main/java/com/reactnativecommunity/viewpager/ReactViewPagerManager.java:144]: I would change it to:

```
viewPager.setOrientation(value.equals("vertical") ? ORIENTATION_VERTICAL : ViewPager2. ORIENTATION_HORIZONTAL);
```

So by default is `horizontal` as it has always been.

File: android/src/main/java/com/reactnativecommunity/viewpager/ReactViewPagerManager.java [hunk 8]
```diff
@@ -1,132 +1,204 @@
...
+        return REACT_CLASS;
+    }
+
+    @NonNull
+    @Override
+    protected ViewPager2 createViewInstance(@NonNull ThemedReactContext reactContext) {
+        final ViewPager2 vp = new ViewPager2(reactContext);
+        FragmentAdapter adapter = new FragmentAdapter((FragmentActivity) reactContext.getCurrentActivity());
+        vp.setAdapter(adapter);
+        vp.setOrientation(ORIENTATION_HORIZONTAL);
```
Comment (ferrannp) [inline android/src/main/java/com/reactnativecommunity/viewpager/ReactViewPagerManager.java:62]: If you change `horizontal` by default, you won't need this line.

**Comment by ferrannp** [comment]
Looking awesome @troZee.  I just added a couple of comments.

Moreover, the example crashes when fast refresh triggers (`unFocus` on a null object reference).

**Comment by troZee** [comment]
@troZee removing pages does not work properly. it needs to be fixed

**Comment by Krizzu** [comment]
@ferrannp 
I checked example and fast refresh works fine for me. Maybe it has to do with Emu/Device API combo?

**Comment by devvit** [comment]
@troZee, everything is fine, nice work 😊

**Comment by reanymeany** [comment]
> Hey @troZee
> 
> Got this version working.. it was a typo that was preventing it from working.
> 
> I can verify that a ScrollView in a vertical pager does not work correctly.
> 
> Here is a little component to reproduce the problem:
> 
> ```
> import ViewPager from "@react-native-community/viewpager";
> import React from "react";
> import { ScrollView, StyleSheet, Text, View } from "react-native";
> 
> const styles = StyleSheet.create({
>   viewPager: {
>     flex: 1,
>   },
> });
> 
> const App = () => (
>   <ViewPager orientation="vertical" style={styles.viewPager} initialPage={0}>
>     <View style={{ padding: 50, backgroundColor: "blue" }}>
>       <Text>Vertical pager, go down to see problem</Text>
>     </View>
> 
>     <View style={{ padding: 50, backgroundColor: "red" }}>
>       <Text>ScrollView in vertical pager does not work</Text>
> 
>       <ScrollView>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>       </ScrollView>
>     </View>
>   </ViewPager>
> );
> 
> export default App;
> ```

Hello, 



> Hey @troZee
> 
> Got this version working.. it was a typo that was preventing it from working.
> 
> I can verify that a ScrollView in a vertical pager does not work correctly.
> 
> Here is a little component to reproduce the problem:
> 
> ```
> import ViewPager from "@react-native-community/viewpager";
> import React from "react";
> import { ScrollView, StyleSheet, Text, View } from "react-native";
> 
> const styles = StyleSheet.create({
>   viewPager: {
>     flex: 1,
>   },
> });
> 
> const App = () => (
>   <ViewPager orientation="vertical" style={styles.viewPager} initialPage={0}>
>     <View style={{ padding: 50, backgroundColor: "blue" }}>
>       <Text>Vertical pager, go down to see problem</Text>
>     </View>
> 
>     <View style={{ padding: 50, backgroundColor: "red" }}>
>       <Text>ScrollView in vertical pager does not work</Text>
> 
>       <ScrollView>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>         <Text style={{ color: "#666", fontSize: 60, fontWeight: "bold" }}>
>           Line
>         </Text>
>       </ScrollView>
>     </View>
>   </ViewPager>
> );
> 
> export default App;
> ```

Hi Miguel,  do you mind sharing where are the typo? i'm having the same error "Could not find com.github.troZee:ViewPager2:v1.0.6." .

**Pull request merged**